src/bit_stream_reader/es_reader.cpp

```diff
@@ -0,0 +1,1163 @@
+        if (end_of_file_) {
+            return total_read_size;
+        }
+        free_space -= read_size;
+        if (free_space == 0) {
+            return total_read_size;
+        }
+    }
+        
+    // Continue filling the beginning part of the ring
+    if (read_ptr_ > 0) {
+        read_size = p_stream_file_.read(reinterpret_cast<char*>(&bs_ring_[write_ptr_]), free_space).gcount();
+        if (read_size > 0) {
+            write_ptr_ = (write_ptr_ + read_size) % BS_RING_SIZE;
+        }
+        if (read_size < free_space) {
+            end_of_file_ = true;
+        }
+        total_read_size += read_size;
+    }
+    return total_read_size;
+}
+
+bool RocVideoESParser::GetByte(int offset, uint8_t *data) {
+    offset = offset % BS_RING_SIZE;
+    if (offset == write_ptr_) {
+        if (FetchBitStream() == 0) {
+            end_of_stream_ = true;
+            return false;
+        }
+    }
+    *data = bs_ring_[offset];
+    return true;
+}
+
+bool RocVideoESParser::ReadBytes(int offset, int size, uint8_t *data) {
+    offset = offset % BS_RING_SIZE;
+    if (size > GetDataSizeInRB()) {
+        if (FetchBitStream() == 0) {
+            end_of_stream_ = true;
+            return false;
+        }
+        if (size > GetDataSizeInRB()) {
+            ERR("Could not read the requested bytes from ring buffer. Either ring buffer size is too small or not enough bytes left.");
+            return false;
+        }
+    }
+    if (offset + size > BS_RING_SIZE) {
+        int part = BS_RING_SIZE - offset;
+        memcpy(data, &bs_ring_[offset], part);
+        memcpy(&data[part], &bs_ring_[0], size - part);
+    } else {
+        memcpy(data, &bs_ring_[offset], size);
+    }
+    return true;
+}
+
+void RocVideoESParser::SetReadPointer(int value) {
+    read_ptr_ = value % BS_RING_SIZE;
+}
+
+bool RocVideoESParser::FindStartCode() {
+    uint8_t three_bytes[3];
+    int i;
+
+    curr_start_code_offset_ = next_start_code_offset_;
+
+    // Search for the next start code
+    while (!end_of_stream_) {
+        for (i = 0; i < 3; i++) {
+            if (GetByte(curr_byte_offset_ + i, three_bytes + i) == false) {
+                break;
+            }
+        }
+        if (i < 3) {
+            break;
+        }
+
+        if (three_bytes[0] == 0 && three_bytes[1] == 0 && three_bytes[2] == 0x01) {
+            num_start_code_++;
+            next_start_code_offset_ = curr_byte_offset_;
+            // Move the pointer 3 bytes forward
+            curr_byte_offset_ = (curr_byte_offset_ + 3) % BS_RING_SIZE;
+
+            // For the very first NAL unit, search for the next start code (or reach the end of frame)
+            if (num_start_code_ == 1) {
+                curr_start_code_offset_ = next_start_code_offset_;
+                continue;
+            } else {
+                break;
+            }
+        }
+        curr_byte_offset_ = (curr_byte_offset_ + 1) % BS_RING_SIZE;
+    }
+    if (num_start_code_ == 0) {
```

Comment:
we can combine the if else to a single statement return (num_start_code_ != 0);

Reply:
Done.